action.yml

```diff
@@ -36,6 +36,14 @@ inputs:
   settings-path:
     description: 'Path to where the settings.xml file will be written. Default is ~/.m2.'
     required: false
+  gpg-private-key:
+    description: 'GPG private key to import. Default is empty string.'
+    required: false
+  gpg-passphrase:
+    description: 'Environment variable name for the GPG private key passphrase. Default is
+       $GPG_PASSPHRASE.'
+    required: false
 runs:
   using: 'node12'
-  main: 'dist/index.js'
+  main: 'dist/setup/index.js'
+  post: 'dist/cleanup/index.js'
```

Comment:
My biggest concern right now is the post cleanup. Is this absolutely necessary? 

If users don't use any of the maven functionality and they just want a version of java to be downloaded and set up, they'll still get the cleanup job to show up in their logs: 

![image](https://user-images.githubusercontent.com/16109154/82575640-07afad80-9b89-11ea-8bae-1f50683b16f6.png)

From the run you linked: https://github.com/jaredpetersen/setup-java-gpg/actions/runs/106387705

In the scenario where a GitHub hosted environment is used, the VM gets destroyed at the very end, so from a security perspective it doesn't make much sense got have to call `gpg.deleteKey`. Is there something I might be missing?

Reply:
The post cleanup is necessary for the shared environment. In the case where people are hosting their own GitHub runners, we don't want want keys kicking around in a GPG keychain after the job completes. They don't have the luxury of a self-cleaning runner.

Reply:
If there's a way to only show that post job cleanup if a user configures GPG, I'd be happy to add it. But as far as I'm aware, the `action.yml` only allows you to run it based on job completion status (like `success`) or at all times.

Reply:
I see that there's a post-if doc [here](https://help.github.com/en/actions/creating-actions/metadata-syntax-for-github-actions#post-if) but I don't believe I can specify that based on an input.

Reply:
So I did some local testing, the `RUNNER_TEMP` directory that is used as part of `getTempDir` gets cleanup up and deleted after each workflow run so it should be safe to store any GPG files. Any of the extra post-cleanup steps could then be avoided.

We have an internal issue open to publicly document `RUNNER_TEMP` and what it actually does, but this seems like it it should do the trick. I'll follow up with some of the folks who are more familiar with `RUNNER_TEMP` and ask if this is safe to use from a security perspective.

Reply:
I double checked with @TingluoHuang and the `RUNNER_TEMP` env variable will bet set for both self-hosted and hosted.

The only time the `RUNNER_TEMP` directory will not be cleaned up at the end is if the job doesn't finish (runner crashes or the machine shuts down). The post job cleanup is susceptible to the same thing so it's safe to just omit all this logic (the whole PR can get considerably simplified now 😃 )

Of course, some mention in the `README` would be good so that users are aware of this behavior and where the GPG file gets created. The temp directory is usually `_work/_temp` and it's in the same location as where the self-hosted runner is running from.

Reply:
The GPG private key is written to the temp directory but the GPG keychain is not stored there. The keychain lives in the user's home directory by default so it won't be cleaned up.

I tried to change the GPG home directory in a previous PR but it doesn't work well on Windows because of the file paths. You have to override the directory separators (\) manually to the posix format in order for it to work. The built in path library does not support this behavior so we have to write that ourselves. Not difficult, but dirty.

Overriding the GPG home directory also mandates that you override the configuration in a new active by default profile in the settings.xml file so that the Maven GPG plugin knows where to look.

This whole thing ends up being dirtier than the post cleanup solution and doesn't allow for any ssh keys that a self hosted runner may already have installed for other purposes in the job.

I think importing the key and removing it results in a cleaner settings.xml file and impacts the runner the least. Just import the key and remove it when you're done. It does involve adding a post job, but I think that's a better trade-off.

Reply:
If you both disagree, let me know and I will implement those changes. I just wanted it to be clear what that entails.

Reply:
😨 That sounds pretty awful. The issue with other ssh keys not working also sounds like a nightmare with lots of potential consequences. This makes me lean towards adding the cleanup step at the end. It's not ideal because the majority of users won't need it, but I think it's the lesser of the two evils. 

This post cleanup job could also have some other use cases in the future, so maybe this isn't all that bad. I'm going to ask a few more people internally to see if there are any major objections to this.

Reply:
> It's not ideal because the majority of users won't need it, but I think it's the lesser of the two evils

That's exactly where I'm at at this point. At least the cleanup step is very quick.

> I'm also going to check if there might be some secret way of running the post cleanup only if it's a self-hosted runner

That would be awesome and is something I didn't even think about! If it's a hosted runner, we don't have to clean anything up because the entire runner gets thrown away.

Reply:
Sorry to rush, but any update?

Reply:
Sorry for the delay, but I'm still waiting on an update so nothing to share at this point (I would really like to get this in though).